### ISSUE TITLE
Document/debug C++ common API usage mistakes

### ISSUE THREAD

**Issue by vjpai** [grpc/grpc#13019]
Please answer these questions before submitting your issue. 
 
### Should this be an issue in the gRPC issue tracker?
 
Yes

### What version of gRPC and what language are you using?
 
Master, C++
 
### What operating system (Linux, Windows, …) and version?
 
N/A
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 
 N/A

### What did you do?

Misused the API
 
### What did you expect to see?

Some info of how I misused it 
 
### What did you see instead?
 
Some kind of crash or other undefined behavior

**Comment by vjpai** [comment]
Common pitfalls and possible solutions/ways to give guidance (please feel free to add comments with more):

1. Multiple outstanding Read or Write on the same stream
    - Add debugging asserts on operation invocation
1. Expecting a result to be available on an async unary or streaming RPC without having called CQ::Next
    - ???
1. Allowing an RPC object to be destructed before all CompletionQueue::Next operations finish
    - Add debugging asserts at destruction

**Comment by wjywbs** [comment]
Hello Vijay, can I use a shared completion queue to send async rpc calls in multiple threads? There will also be one more thread that loops cq.Next().

**Comment by vjpai** [comment]
Yes, the CQ is absolutely thread safe! If you use multi-threaded CQ from an async client, I strongly suggest you use the Prepare... form of the stub interface that was implemented in #12269 .This requires 2 calls to start an async RPC, but it helps to prevent a race (or alternatively, resolves a workaround that requires locking). I will add this to the list.

**Comment by vjpai** [comment]
Cc: @slash-lib @jeady (soliciting opinions)

**Comment by slash-lib** [comment]
Looks good - except 5: isn't the protobuf serialized serially (!) before Write returns?

**Comment by ctiller** [comment]
Currently true but we might want to make it less true (serialized before write completes) and push serialization down our stack... The advantage being we can get better cache coherency and we don't need to hold the entire serialized message in memory at once.

**Comment by slash-lib** [comment]
I see your point, but it means IMO the API is incorrect: Write should take a const*, not a const& - it's a common idiom that const& does not need to outlive the function call.

**Comment by vjpai** [comment]
I should add something about "expecting any kind of progress on RPCs, even for write-related activities without a CQ Next somewhere" as well as "expecting writes to complete without a matching read on the other side of the stream"; both are common mistakes since for certain message sizes and transports, writes can be completed by buffering into flow-control, but that's not always possible.
